### PR TITLE
Context and decorator for module_train and requires_grad

### PR DESCRIPTION
**Pull request by samedii**

### CODE REVIEW

File: workflow/torch/requires_grad.py [hunk 1]
```diff
@@ -1,29 +1,32 @@
...
+    def __init__(self, module, grad=True):
+        self.module = module
+        self.grad = grad
+        self.requires_grads = [
+            p.requires_grad for p in module.parameters()
+        ]
 
-    yield module
+    def __enter__(self):
+        for p in self.module.parameters():
```
Comment (FelixAbrahamsson) [inline workflow/torch/requires_grad.py:14]: Will this not mess up parameters that originally didn't require grad? 
Don't we need to do
```
for p, requires_grad in zip(self.module.parameters(), self.requires_grads):
    p.requires_grad = self.grad and requires_grad
```

Comment (FelixAbrahamsson) [reply workflow/torch/requires_grad.py:14]: Example
```
>>> class MyModel(nn.Module):
>>>     def __init__(self):
>>>         super().__init__()
>>>         self.param = nn.Parameter(torch.randn(1), requires_grad=False)
>>>         self.dense = nn.Linear(1, 1)
>>>     def forward(self, x):
>>>         return self.dense(x) + self.param

>>> model = MyModel()
>>> for p in model.parameters():
>>>     print(p.requires_grad)
False
True
True

>>> with requires_grad(model):
>>>     for p in model.parameters():
>>>         print(p.requires_grad)
True
True
True
```
Shouldn't be desired behaviour?

Comment (FelixAbrahamsson) [inline workflow/torch/requires_grad.py:14]: How do we handle parameters that are never supposed to require grad?
Example of undesired behavior:
```
>>> class MyModel(nn.Module):
>>>     def __init__(self):
>>>         super().__init__()
>>>         self.param = nn.Parameter(torch.randn(1), requires_grad=False)
>>>         self.dense = nn.Linear(1, 1)
>>>     def forward(self, x):
>>>         return self.dense(x) + self.param

>>> model = MyModel()

>>> for p in model.parameters():
>>>     print(p.requires_grad)
False
True
True

>>> with requires_grad(model):
>>>     for p in model.parameters():
>>>         print(p.requires_grad)
True
True
True
```

File: workflow/torch/requires_grad.py [hunk 2]
```diff
@@ -1,29 +1,32 @@
...
+            p.requires_grad = self.grad
```
Comment (FelixAbrahamsson) [inline workflow/torch/requires_grad.py:15]: How do we handle parameters that are never supposed to require grad?
Example of undesired behavior:
```
>>> class MyModel(nn.Module):
>>>     def __init__(self):
>>>         super().__init__()
>>>         self.param = nn.Parameter(torch.randn(1), requires_grad=False)
>>>         self.dense = nn.Linear(1, 1)
>>>     def forward(self, x):
>>>         return self.dense(x) + self.param

>>> model = MyModel()

>>> for p in model.parameters():
>>>     print(p.requires_grad)
False
True
True

>>> with requires_grad(model):
>>>     for p in model.parameters():
>>>         print(p.requires_grad)
True
True
True
```

Comment (samedii) [reply workflow/torch/requires_grad.py:15]: Very good point. I haven't considered this. Maybe we should not use nn.Parameter for everything?

Comment (samedii) [reply workflow/torch/requires_grad.py:15]: It seems that we should use `register_buffer` for things that should not ever have grads: https://pytorch.org/docs/stable/generated/torch.nn.Module.html?highlight=register_buffer#torch.nn.Module.register_buffer

Comment (FelixAbrahamsson) [reply workflow/torch/requires_grad.py:15]: That looks like the correct way to do it!

**Review by FelixAbrahamsson — APPROVED** [review]

**Pull request merged**